src/core/namespace.c

```diff
@@ -3310,6 +3313,379 @@ bool ns_type_supported(NamespaceType type) {
+                        r = move_mount(hierarchy_path_fd, "", AT_FDCWD, hierarchy_path_moved_mount, MOVE_MOUNT_F_EMPTY_PATH);
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to move mount for %s: %m", mount_path);
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        /* Turn all overlay layer directories into FD-based references */
+                        STRV_FOREACH(ol, m->overlay_layers) {
+                                int chased_fd;
+
+                                chased_fd = open_tree(-EBADF, *ol, 0);
+                                if (chased_fd < 0) {
+                                        log_debug_errno(errno, "Failed to open_tree overlay layer: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+
+                                r = strv_push(&new_layers, FORMAT_PROC_FD_PATH(chased_fd)); // todo path join
+                                if (r < 0) {
+                                        log_debug_errno(r, "Failed to append overlay layer: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+                        }
+                        m->overlay_layers = strv_free(m->overlay_layers);
+                        m->overlay_layers = TAKE_PTR(new_layers);
+
+                        layers = strv_join(m->overlay_layers, ":");
+                        if (!layers)
+                                _exit(EXIT_FAILURE);
+
+                        /* Append the underlying hierarchy path as the last lowerdir */
+                        options = strjoin(layers, ":", FORMAT_PROC_FD_PATH(hierarchy_path_fd));
+                        if (!options)
+                                _exit(EXIT_FAILURE);
+
+                        r = fsconfig(overlay_fs_fd, FSCONFIG_SET_STRING, "lowerdir", options, 0);
+                        if (r < 0) {
+                                log_debug_errno(SYNTHETIC_ERRNO(errno), "Failed to set lowerdir: %m");
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        /* Create the superblock */
+                        r = fsconfig(overlay_fs_fd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to create overlay superblock: %m");
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        /* Signal completion to grandchild */
+                        r = send_one_fd(pair[0], overlay_fs_fd, 0);
+                        if (r < 0) {
+                                log_debug_errno(r, "Failed to signal overlay configuration complete for %s: %m", mount_path);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
```

Comment:
so you moved the important part of the grandchild into its own helper func unpeel_mount_and_setup_overlay(). I think it would make sense to move this part here – as the other side of the socketpair protocol – into its own function too, that is somewhat symmetrical in the way it looks, so that the protocol interactions are reasonably easy to grasp.